src/mlpack/tests/lmnn_test.cpp

```diff
@@ -442,53 +443,171 @@ BOOST_AUTO_TEST_CASE(LMNNAccuracyTest)
   BOOST_REQUIRE_CLOSE(finalAccuracy, 100.0, 1e-5);
 }
 
-// Check that accuracy while learning square distance matrix is same
-// as when we are learnig low rank matrix.
+// Check that accuracy while learning square distance matrix is the same as when
+// we are learning low rank matrix.  I'm ok if this passes only once out of
+// three tries.
+BOOST_AUTO_TEST_CASE(LMNNLowRankAccuracyLBFGSTest)
+{
+  bool success = false;
+  for (size_t trial = 0; trial < 3; ++trial)
+  {
+    arma::mat dataPart1;
+    dataPart1.randn(5, 50);
+
+    arma::Row<size_t> labelsPart1(50);
+    labelsPart1.fill(0);
+
+    arma::mat dataPart2;
+    dataPart2.randn(5, 50);
+
+    arma::Row<size_t> labelsPart2(50);
+    labelsPart2.fill(1);
+
+    // Generate ordering.
+    arma::uvec ordering = arma::shuffle(arma::linspace<arma::uvec>(0, 99, 100));
+
+    // Generate datasets.
+    arma::mat dataset = join_rows(dataPart1, dataPart2);
+    dataset = dataset.cols(ordering);
+
+    // Generate labels.
+    arma::Row<size_t> labels = join_rows(labelsPart1, labelsPart2);
+    labels = labels.cols(ordering);
+
+    LMNN<SquaredEuclideanDistance, L_BFGS> lmnn(dataset, labels, 1);
+
+    // Learn a square matrix.
+    arma::mat outputMatrix;
+    lmnn.LearnDistance(outputMatrix);
+
+    double acc1 = KnnAccuracy(outputMatrix * dataset, labels, 1);
+
+    // Learn a low rank matrix.
+    outputMatrix = arma::randu(4, 5);
+    lmnn.LearnDistance(outputMatrix);
+
+    double acc2 = KnnAccuracy(outputMatrix * dataset, labels, 1);
+
+    // We keep the tolerance very high.  We need to ensure the accuracy drop
+    // isn't any more than 10%.
+    success = ((acc1 - acc2) <= 10.0);
```

Comment:
Why we are only considering  accuracy drop case? I think it accuracy may increase as well and in that case this will always result in success no matter how much the difference. So, shouldn't we use `abs(acc1 - acc2)` here. What do you think?

Reply:
If accuracy increases, even by a lot, I don't see this as an issue at all.  Mostly I want to check that accuracy doesn't drop way off, because to me that could indicate a big bug.  (Like if accuracy goes from 50% to 0%, for instance.)

Reply:
hmm, Then it's all right. I think it's good to merge :)